### PR TITLE
PLNSRVCE-528 fix issue with contaminated builds

### PR DESCRIPTION
**Pull request by stuartwdouglas**

### CODE REVIEW

**Comment by gabemontero** [comment]
Is there something present on a artifactbuild or dependencybuild CRD instance as a result of this fix that we could cross-verify against in the in-repo e2e (assuming the repo we settle on for our in-repo tests has a contaminant) ?

 I *think* https://github.com/stuartwdouglas/code-with-quarkus has a contaminant if we let it run long enough. But whether we can tolerate https://github.com/stuartwdouglas/code-with-quarkus in our PR e2e's long term because of resource constraints is TBD (I believe I've seen intermittent instances of resources constraints during maven download in a case or two.

When I start trying out https://github.com/stuartwdouglas/shaded-java-app today I'll see if it has one (though you quite possibly know that off the top of your head @stuartwdouglas )

/retest

for pipeline as code failures

**Comment by gabemontero** [comment]
> Is there something present on a artifactbuild or dependencybuild CRD instance as a result of this fix that we could cross-verify against in the in-repo e2e (assuming the repo we settle on for our in-repo tests has a contaminant) ?
> 
> I _think_ https://github.com/stuartwdouglas/code-with-quarkus has a contaminant if we let it run long enough. But whether we can tolerate https://github.com/stuartwdouglas/code-with-quarkus in our PR e2e's long term because of resource constraints is TBD (I believe I've seen intermittent instances of resources constraints during maven download in a case or two.)
> 
> When I start trying out https://github.com/stuartwdouglas/shaded-java-app today I'll see if it has one (though you quite possibly know that off the top of your head @stuartwdouglas )
> 
> /retest
> 
> for pipeline as code failures

yeah no contaminated builds with https://github.com/stuartwdouglas/shaded-java-app 
which I think we want to switch to because I am still seeing intermittent resource constraints with the quarkus repo

**Comment by gabemontero** [comment]
> > Is there something present on a artifactbuild or dependencybuild CRD instance as a result of this fix that we could cross-verify against in the in-repo e2e (assuming the repo we settle on for our in-repo tests has a contaminant) ?
> > I _think_ https://github.com/stuartwdouglas/code-with-quarkus has a contaminant if we let it run long enough. But whether we can tolerate https://github.com/stuartwdouglas/code-with-quarkus in our PR e2e's long term because of resource constraints is TBD (I believe I've seen intermittent instances of resources constraints during maven download in a case or two.)
> > When I start trying out https://github.com/stuartwdouglas/shaded-java-app today I'll see if it has one (though you quite possibly know that off the top of your head @stuartwdouglas )
> > /retest
> > for pipeline as code failures
> 
> yeah no contaminated builds with https://github.com/stuartwdouglas/shaded-java-app which I think we want to switch to because I am still seeing intermittent resource constraints with the quarkus repo

Correction - one contaminated 

`dependencybuild.jvmbuildservice.io/d1f3a6d34e891ddc731c33f45f8e05de   https://github.com/netty/netty.git              netty-4.1.74.Final                   DependencyBuildStateContaminated   
`

Though looking at its yaml, I don't see how we could cross ref with the changes here.

/approve
/lgtm

**Comment by stuartwdouglas** [comment]
@gabemontero we should verify that the jctools ABR gets created, and then that Netty rebuilds.

If you want to test the end state I think we really need to verify that Netty is complete, JCTools is complete, and Netty started after the JCTools build.

**Comment by stuartwdouglas** [comment]
@gabemontero I have had an attempted at adding an E2E test for this flow.

File: openshift-with-appstudio-test/e2e/jvm_build_service_test.go [hunk 1]
```diff
@@ -395,4 +396,57 @@ func TestExampleRun(t *testing.T) {
...
 		}
 	})
+
+	ta.t.Run("contaminated build is resolved", func(t *testing.T) {
+		//our sample repo has Netty which is contaminated by JCTools
+		var contaminated string
+		err = wait.PollImmediate(ta.interval, 2*ta.timeout, func() (done bool, err error) {
+
+			dbContaminated := false
+			dbList, err := jvmClient.JvmbuildserviceV1alpha1().DependencyBuilds(ta.ns).List(context.TODO(), metav1.ListOptions{})
```
Comment (gabemontero) [inline openshift-with-appstudio-test/e2e/jvm_build_service_test.go:406]: log this error if it is not nil ... might help us in resource constraint situations

**Review by gabemontero — CHANGES_REQUESTED** [review]
just a couple of minor suggestions @stuartwdouglas 

I've seen enough odd errors from the api server over the years ... reporting them in this case always proves helpful in those rare / intermittent cases

**Comment by stuartwdouglas** [comment]
Done

**Review by gabemontero — APPROVED** [review]
cool let's see how it does

**Comment by stuartwdouglas** [comment]
The failure is unrelated but I have pushed a fix to this branch

**Comment by stuartwdouglas** [comment]
Somewhat worryingly it looks like the sidecar has just randomly stopped accepting connections, then is still attempting to do IO after the build is over. I am really not sure what is happening there.

**Comment by gabemontero** [comment]
> Somewhat worryingly it looks like the sidecar has just randomly stopped accepting connections, then is still attempting to do IO after the build is over. I am really not sure what is happening there.

Ok @stuartwdouglas coming in this AM and looking here, the e2e is green.

Thas said, unless you already this and feel otherwise, I can bring up an openshift cluster today, check out your PR, and retest locally, and see if I see failures, or see the sidecar anomaly your are describing and try to debug.  I'll post any findings here, and we can reconvene later my Wednesday / your Thursday and decide what we want to do.

**Comment by stuartwdouglas** [comment]
I don't think it is related in any way, just something I would like to understand, it could be a problem with the sidecar, or something environmental.

**Comment by gabemontero** [comment]
> I don't think it is related in any way, just something I would like to understand, it could be a problem with the sidecar, or something environmental.

Cool deal.  If I have cycles I might try today with this merged if I can reproduced / identify it, for educational purposes if anything else.

**Comment by gabemontero** [comment]
> > I don't think it is related in any way, just something I would like to understand, it could be a problem with the sidecar, or something environmental.
> 
> Cool deal. If I have cycles I might try today with this merged if I can reproduced / identify it, for educational purposes if anything else.

So yeah I discovered first hand the debug difficulties sidecars provide ;-)

in the end, no discoveries as to what if anything is up with the sidecar or env, but it did lead to some general test robustness tweaks, as well as *potential* sidecar debug updates I'll note to you in my catch all for today, https://github.com/redhat-appstudio/jvm-build-service/pull/135

**Pull request merged**